src/collection.rs

```diff
@@ -429,52 +476,354 @@ impl<'a, C: ClientExt> Collection<'a, C> {
+    /// contains three special attributes: _id containing the document
+    /// identifier, _key containing key which uniquely identifies a document in
+    /// a given collection and _rev containing the revision.
+    #[maybe_async]
+    pub async fn read_document<T>(&self, _key: &str) -> Result<Document<T>, ClientError>
+    where
+        T: Serialize + Debug + DeserializeOwned,
+    {
+        self.read_document_with_options(_key, None).await
+    }
+
+    #[maybe_async]
+    pub async fn read_document_with_options<T>(
+        &self,
+        _key: &str,
+        read_options: Option<DocumentReadOptions>,
+    ) -> Result<Document<T>, ClientError>
+    where
+        T: Serialize + Debug + DeserializeOwned,
+    {
+        let url = self.document_base_url.join(_key).unwrap();
+        let mut build = Request::get(url.to_string());
+
+        let header = make_header_from_options(read_options);
+        if let Some(h) = header {
+            build = build.header(h.0.as_str(), h.1.as_str())
+        }
+        let req = build.body("".to_string()).unwrap();
+        let resp: Document<T> = deserialize_response(self.session.request(req).await?.body())?;
+        Ok(resp)
+    }
+
+    /// Reads a single document header
+    /// Like GET, but only returns the header fields and not the body. You can
+    /// use this call to get the current revision of a document or check if the
+    /// document was deleted.
+    #[maybe_async]
+    pub async fn read_document_header(&self, _key: &str) -> Result<DocumentHeader, ClientError> {
+        self.read_document_header_with_options(_key, None).await
+    }
+
+    #[maybe_async]
+    pub async fn read_document_header_with_options(
+        &self,
+        _key: &str,
+        read_options: Option<DocumentReadOptions>,
+    ) -> Result<DocumentHeader, ClientError>
+where {
+        let url = self.document_base_url.join(_key).unwrap();
+        let mut build = Request::get(url.to_string());
+
+        let header = make_header_from_options(read_options);
+        if let Some(h) = header {
+            build = build.header(h.0.as_str(), h.1.as_str())
+        }
+        let req = build.body("".to_string()).unwrap();
+        let resp: DocumentHeader = deserialize_response(self.session.request(req).await?.body())?;
+        Ok(resp)
+    }
     /// Partially updates the document
-    ///
-    /// # Note
-    /// this function would make a request to arango server.
     #[maybe_async]
-    pub async fn update_document<T>(&self, _doc: Document<T>) {
-        unimplemented!()
+    pub async fn update_document<T>(
+        &self,
+        _key: &str,
+        doc: T,
+        update_options: Option<DocumentUpdateOptions>,
+    ) -> Result<DocumentResponse<T>, ClientError>
+    where
+        T: Serialize + DeserializeOwned,
+    {
+        let mut url = self.document_base_url.join(_key).unwrap();
+        let body = serde_json::to_string(&doc)?;
+        if let Some(options) = update_options {
```

Comment:
This pattern to match every query is quite annoying. Can we use a library to help serialize into query string, e.x. [serde_qs](https://docs.rs/serde_qs/0.6.1/serde_qs/)?